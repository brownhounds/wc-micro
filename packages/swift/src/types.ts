export type TestType = {
    something: 'hello';
};
