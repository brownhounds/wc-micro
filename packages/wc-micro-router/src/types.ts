export type ComponentDefinition = {
    tag: string;
    import: () => Promise<unknown>;
};

export type RouterConfig = {
    routerTag?: string;
    outletTag?: string;
    routePrefix?: string;
    notFound?: ComponentDefinition;
};

export type RouteObject = {
    path: string;
    outletId?: string;
    children?: RouteObject[];
    topLevelRoute?: RouteObject; // TODO: This should not be exposed to a user!!
    component: ComponentDefinition;
};

export type HydratedRoute = {
    route: string;
    pathname: string;
    params: Record<string, string>;
};

export const RouterEvents = {
    NAVIGATE: 'router-navigate',
} as const;

export type RouterEventsType = (typeof RouterEvents)[keyof typeof RouterEvents];

export type NavigateConfig = { replace: boolean };

export type RouterContextConfig = {
    routerTag: string;
    outletTag: string;
    routePrefix: string;
    notFound?: ComponentDefinition;
};
