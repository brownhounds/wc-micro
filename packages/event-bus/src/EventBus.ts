export type ListenerCallback<Payload> = (payload?: Payload) => void;

export class EventBus<EventType, Payload> {
    #listeners = new Map<EventType, Set<ListenerCallback<Payload>>>();

    public get listeners(): Map<EventType, Set<ListenerCallback<Payload>>> {
        return this.#listeners;
    }

    public addListener(
        eventType: EventType,
        listenerCallback: ListenerCallback<Payload>
    ): void {
        if (!this.#listeners.get(eventType)) {
            this.#listeners.set(eventType, new Set([listenerCallback]));
            return;
        }

        this.#listeners.get(eventType)?.add(listenerCallback);
    }

    public removeListener(
        eventType: EventType,
        listenerCallback: ListenerCallback<Payload>
    ): void {
        const listeners = this.#listeners.get(eventType);
        if (listeners && listeners.size > 1) {
            listeners?.delete(listenerCallback);
            return;
        }

        this.#listeners.delete(eventType);
    }

    public dispatch(eventType: EventType, payload?: Payload): void {
        const listeners = this.#listeners.get(eventType);
        if (listeners) {
            for (const listener of listeners) {
                listener(payload);
            }
        }
    }
}
