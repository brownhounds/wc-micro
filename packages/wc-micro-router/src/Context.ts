import { EventBus } from '@brownhounds/event-bus';

import type {
    NavigateConfig,
    RouteObject,
    RouterConfig,
    RouterContextConfig,
    RouterEventsType,
} from './types';

export class RouterContext {
    public static config: RouterContextConfig = {
        routerTag: 'router-element',
        outletTag: 'outlet-element',
        routePrefix: '',
    };
    public static definedRoutes: RouteObject[];
    public static routes: RouteObject[] = [];
    public static eventBus = new EventBus<RouterEventsType, RouteObject>();
    public static navigate: (
        pathname: string,
        config?: NavigateConfig,
        popState?: boolean
    ) => void;
    public static navigateLocation: (pathname: string) => void;
    public static popState: (pathname: string) => void;

    public static mergeConfigs(config: RouterConfig): void {
        RouterContext.config = {
            ...RouterContext.config,
            ...config,
        };
    }

    public static handleError(error: Error): never {
        throw error;
    }
}
