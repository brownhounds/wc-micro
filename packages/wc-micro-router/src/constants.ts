import type { ComponentDefinition } from './types';

export const NOT_FOUND_TAG = 'router-not-found';

export const builtInNotFound: ComponentDefinition = {
    tag: NOT_FOUND_TAG,
    import: () => import('./HtmlElements/not-found-element'),
};
