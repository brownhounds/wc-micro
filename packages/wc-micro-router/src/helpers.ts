import { RouterContext } from './Context';
import type { RouteObject } from './types';

export const formatRoute = (route: string): string => {
    return (
        '/' +
        route
            .split('/')
            .filter((part) => part !== '')
            .join('/')
    );
};

const findDuplicatedValuesInArray = (array: string[]): string[] =>
    array.filter((e, i, a) => a.indexOf(e) !== i);

const findDuplicatedRoutes = (array: RouteObject[]): Set<string> => {
    const seen = new Set<string>();
    const duplicates = new Set<string>();

    for (const route of array) {
        if (seen.has(route.path)) duplicates.add(route.path);
        seen.add(route.path);
    }

    return duplicates;
};

const validateChildrenOutletIds = (routes: RouteObject[]): void => {
    const invalidRoute = routes.find((route) => !route.outletId);
    if (invalidRoute) {
        RouterContext.handleError(
            new Error(`Child Route: ${invalidRoute.path} has missing outletId`)
        );
    }
};

const validateUniqueRoutes = (): void => {
    const duplicates = findDuplicatedRoutes(RouterContext.routes);
    if (duplicates.size) {
        RouterContext.handleError(
            new Error(
                `Found duplicated routes: ${Array.from(duplicates).join(', ')}`
            )
        );
    }
};

const matchParamNames = (pattern: string): string[] =>
    pattern.match(/:[^/]+/g) || [];

const validateUniqueRouteParams = (): void => {
    for (const route of RouterContext.routes) {
        const duplicates = findDuplicatedValuesInArray(
            matchParamNames(route.path)
        );
        if (duplicates.length) {
            RouterContext.handleError(
                new Error(
                    `Route: ${route.path} has duplicated dynamic parameters ${duplicates.join(', ')}`
                )
            );
        }
    }
};

const matchParams = (
    pattern: string,
    pathname: string
): { paramNames: string[]; match: RegExpMatchArray } => {
    const regex = new RegExp(pattern.replace(/:[^/]+/g, '([^/]+)'));
    const match = pathname.match(regex);

    return {
        paramNames: matchParamNames(pattern),
        match: match as RegExpMatchArray,
    };
};

export const matchRoute = (pattern: string, pathname: string): boolean => {
    const regexPattern = pattern
        .replace(/:[a-zA-Z_]+/g, '([\\w-]+)')
        .replace(/\//g, '\\/');
    const regex = new RegExp(`^${regexPattern}$`);
    const match = pathname.match(regex);
    return match ? true : false;
};

export const castParamsToObject = (
    pattern: string,
    pathname: string
): Record<string, string> => {
    const { paramNames, match } = matchParams(pattern, pathname);
    const params: Record<string, string> = {};

    for (let i = 0; i < paramNames.length; i++) {
        const paramName = paramNames[i].slice(1);
        params[paramName] = match[i + 1];
    }

    return params;
};

const compileChildRoutes = (
    routes: RouteObject[],
    topLevelRoute: RouteObject,
    parentRoutePath: string = ''
): void => {
    for (const route of routes) {
        RouterContext.routes.push({
            ...route,
            path: formatRoute(
                RouterContext.config.routePrefix +
                    topLevelRoute.path +
                    parentRoutePath +
                    route.path
            ),
            topLevelRoute,
        });

        if (route.children) {
            validateChildrenOutletIds(route.children);
            compileChildRoutes(route.children, topLevelRoute, route.path);
        }
    }
};

export const compileRoutes = (): void => {
    for (const route of RouterContext.definedRoutes) {
        RouterContext.routes.push({
            ...route,
            path: formatRoute(RouterContext.config.routePrefix + route.path),
        });
        if (route.children?.length) {
            validateChildrenOutletIds(route.children);
            compileChildRoutes(route.children, route);
        }
    }

    validateUniqueRoutes();
    validateUniqueRouteParams();
};
