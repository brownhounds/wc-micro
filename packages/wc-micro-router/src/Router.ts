import { RouterContext } from './Context';
import { castParamsToObject, formatRoute, matchRoute } from './helpers';
import type { RouteObject } from './types';
import { RouterEvents, type NavigateConfig, type HydratedRoute } from './types';

const EMPTY_ROUTE = { route: '', pathname: '', params: {} };

export class Router {
    #currentRoute: HydratedRoute = EMPTY_ROUTE;

    public get activeRoute(): HydratedRoute {
        return this.#currentRoute;
    }

    #matchRoute(pathname: string): RouteObject | undefined {
        for (const route of RouterContext.routes) {
            if (matchRoute(route.path, pathname)) {
                return route;
            }
        }

        return undefined;
    }

    #hydrateActiveRoute(route: RouteObject, pathname: string): HydratedRoute {
        return {
            route: route.path,
            pathname,
            params: castParamsToObject(route.path, pathname),
        };
    }

    #setActiveRoute(pathname: string): RouteObject | undefined {
        const matchedRoute = this.#matchRoute(pathname);
        if (matchedRoute) {
            this.#currentRoute = this.#hydrateActiveRoute(
                matchedRoute,
                pathname
            );
        } else {
            this.#currentRoute = EMPTY_ROUTE;
        }

        return matchedRoute;
    }

    #dispatchNavigationEvent(matchedRoute?: RouteObject): void {
        RouterContext.eventBus.dispatch(RouterEvents.NAVIGATE, matchedRoute);
    }

    public navigate(pathname: string, config?: NavigateConfig): void {
        const formattedPathname = formatRoute(
            RouterContext.config.routePrefix + formatRoute(pathname)
        );
        const matchedRoute = this.#setActiveRoute(formattedPathname);
        config?.replace
            ? window.history.replaceState({}, '', formattedPathname)
            : window.history.pushState({}, '', formattedPathname);
        this.#dispatchNavigationEvent(matchedRoute);
    }

    // TODO: What do I do with this, I don't like it!!
    public navigateLocation(pathname: string): void {
        const formattedPathname = formatRoute(pathname);
        const matchedRoute = this.#setActiveRoute(formattedPathname);
        window.history.pushState({}, '', formattedPathname);
        this.#dispatchNavigationEvent(matchedRoute);
    }

    public popState(pathname: string): void {
        const formattedPathname = formatRoute(
            RouterContext.config.routePrefix + formatRoute(pathname)
        );
        const matchedRoute = this.#setActiveRoute(formattedPathname);
        this.#dispatchNavigationEvent(matchedRoute);
    }

    public matchRoute(pattern: string): boolean {
        const patternWithPrefix = RouterContext.config.routePrefix + pattern;
        const route = RouterContext.routes.find(
            (route) => route.path === patternWithPrefix
        );

        if (matchRoute(patternWithPrefix, window.location.pathname) && route)
            return true;

        return false;
    }
}
