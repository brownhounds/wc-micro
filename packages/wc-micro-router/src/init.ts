import { RouterContext } from './Context';
import { Router } from './Router';
import { compileRoutes } from './helpers';
import type { RouteObject, RouterConfig } from './types';

export const initRouter = (
    routes: RouteObject[],
    config?: RouterConfig
): Router => {
    if (config) RouterContext.mergeConfigs(config);
    RouterContext.definedRoutes = routes;

    compileRoutes();

    const router = new Router();
    RouterContext.navigate = router.navigate.bind(router);
    RouterContext.navigateLocation = router.navigateLocation.bind(router);
    RouterContext.popState = router.popState.bind(router);
    import('./HtmlElements/elements');

    return router;
};
