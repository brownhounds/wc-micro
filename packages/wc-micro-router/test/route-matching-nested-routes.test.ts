import { expect, test, describe, beforeAll, afterAll } from 'vitest';
import { Present } from '@brownhounds/wc-testing';
import { HTML } from '@brownhounds/wc-testing/snapshots';
import { html } from '@brownhounds/uhtml';
import { initRouter } from '../src/init';
import type { Router } from '../src/Router';
import { restoreRouterContext } from './mocks/Router';

describe('Route Matching - Nested Routes', () => {
    let router: Router;

    beforeAll(() => {
        router = initRouter([
            {
                path: '/',
                component: {
                    tag: 'page-one-outlet',
                    import: (): Promise<unknown> =>
                        import('./fixtures/Page1withOutlet'),
                },
                children: [
                    {
                        outletId: 'page-one-nested-route',
                        path: '/page/:pageId',
                        component: {
                            tag: 'page-two-outlet',
                            import: (): Promise<unknown> =>
                                import('./fixtures/Page2withOutlet'),
                        },
                        children: [
                            {
                                outletId: 'page-two-nested-route',
                                path: '/category/:categoryId',
                                component: {
                                    tag: 'page-one',
                                    import: (): Promise<unknown> =>
                                        import('./fixtures/Page1'),
                                },
                            },
                        ],
                    },
                ],
            },
        ]);
    });

    afterAll(() => {
        restoreRouterContext();
    });

    test('Match Route: `/`', async () => {
        window.location.pathname = '/';
        const presenter = new Present().screen(html`<router-element />`);

        await presenter.wait(100);

        expect(HTML(presenter.root())).toMatchSnapshot();

        expect(router.activeRoute).toMatchObject({
            route: '/',
            pathname: '/',
            params: {},
        });
    });

    test('Match Route: `/page/:pageId`', async () => {
        window.location.pathname = '/page/123456';
        const presenter = new Present().screen(html`<router-element />`);

        await presenter.wait(100);

        expect(HTML(presenter.root())).toMatchSnapshot();

        expect(router.activeRoute).toMatchObject({
            route: '/page/:pageId',
            pathname: '/page/123456',
            params: {
                pageId: '123456',
            },
        });
    });

    test('Match Route: `/page/:pageId/category/:categoryId`', async () => {
        window.location.pathname = '/page/123456/category/56987';
        const presenter = new Present().screen(html`<router-element />`);

        await presenter.wait(100);

        expect(HTML(presenter.root())).toMatchSnapshot();

        expect(router.activeRoute).toMatchObject({
            route: '/page/:pageId/category/:categoryId',
            pathname: '/page/123456/category/56987',
            params: {
                pageId: '123456',
                categoryId: '56987',
            },
        });
    });

    test('Navigate', async () => {
        window.location.pathname = '/';
        const presenter = new Present().screen(html`<router-element />`);

        await presenter.wait(100);

        expect(HTML(presenter.root())).toMatchSnapshot();

        expect(router.activeRoute).toMatchObject({
            route: '/',
            pathname: '/',
            params: {},
        });

        router.navigate('/page/123456');

        await presenter.wait(100);

        expect(HTML(presenter.root())).toMatchSnapshot();

        expect(router.activeRoute).toMatchObject({
            route: '/page/:pageId',
            pathname: '/page/123456',
            params: {
                pageId: '123456',
            },
        });

        router.navigate('/page/123456/category/56987');
        await presenter.wait(100);

        expect(HTML(presenter.root())).toMatchSnapshot();

        expect(router.activeRoute).toMatchObject({
            route: '/page/:pageId/category/:categoryId',
            pathname: '/page/123456/category/56987',
            params: {
                pageId: '123456',
                categoryId: '56987',
            },
        });

        history.back();

        await presenter.wait(100);

        expect(HTML(presenter.root())).toMatchSnapshot();

        expect(router.activeRoute).toMatchObject({
            route: '/page/:pageId',
            pathname: '/page/123456',
            params: {
                pageId: '123456',
            },
        });
    });
});
