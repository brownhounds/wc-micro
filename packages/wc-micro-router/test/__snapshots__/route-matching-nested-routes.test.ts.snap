// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Route Matching - Nested Routes > Match Route: \`/\` 1`] = `
"
<router-element>
  <page-one-outlet><!-- #shadowroot -->
    <h1>Page 1 - With Outlet</h1>
    <p>/</p>
    <outlet-element id="page-one-nested-route"></outlet-element>
  </page-one-outlet>
</router-element>
"
`;

exports[`Route Matching - Nested Routes > Match Route: \`/page/:pageId/category/:categoryId\` 1`] = `
"
<router-element>
  <page-one-outlet><!-- #shadowroot -->
    <h1>Page 1 - With Outlet</h1>
    <p>/page/123456/category/56987</p>
    <outlet-element id="page-one-nested-route">
      <page-two-outlet><!-- #shadowroot -->
        <h1>Page 2 - With Outlet</h1>
        <p>/page/123456/category/56987</p>
        <outlet-element id="page-two-nested-route">
          <page-one><!-- #shadowroot -->
            <h1>Page 1</h1>
            <p>/page/123456/category/56987</p>
          </page-one>
        </outlet-element>
      </page-two-outlet>
    </outlet-element>
  </page-one-outlet>
</router-element>
"
`;

exports[`Route Matching - Nested Routes > Match Route: \`/page/:pageId\` 1`] = `
"
<router-element>
  <page-one-outlet><!-- #shadowroot -->
    <h1>Page 1 - With Outlet</h1>
    <p>/page/123456</p>
    <outlet-element id="page-one-nested-route">
      <page-two-outlet><!-- #shadowroot -->
        <h1>Page 2 - With Outlet</h1>
        <p>/page/123456</p>
        <outlet-element id="page-two-nested-route"></outlet-element>
      </page-two-outlet>
    </outlet-element>
  </page-one-outlet>
</router-element>
"
`;

exports[`Route Matching - Nested Routes > Navigate 1`] = `
"
<router-element>
  <page-one-outlet><!-- #shadowroot -->
    <h1>Page 1 - With Outlet</h1>
    <p>/</p>
    <outlet-element id="page-one-nested-route"></outlet-element>
  </page-one-outlet>
</router-element>
"
`;

exports[`Route Matching - Nested Routes > Navigate 2`] = `
"
<router-element>
  <page-one-outlet><!-- #shadowroot -->
    <h1>Page 1 - With Outlet</h1>
    <p>/</p>
    <outlet-element id="page-one-nested-route"></outlet-element>
  </page-one-outlet>
</router-element>
"
`;

exports[`Route Matching - Nested Routes > Navigate 3`] = `
"
<router-element>
  <page-one-outlet><!-- #shadowroot -->
    <h1>Page 1 - With Outlet</h1>
    <p>/</p>
    <outlet-element id="page-one-nested-route"></outlet-element>
  </page-one-outlet>
</router-element>
"
`;

exports[`Route Matching - Nested Routes > Navigate 4`] = `
"
<router-element>
  <page-one-outlet><!-- #shadowroot -->
    <h1>Page 1 - With Outlet</h1>
    <p>/</p>
    <outlet-element id="page-one-nested-route"></outlet-element>
  </page-one-outlet>
</router-element>
"
`;
